src/services/refactors/extractMethod.ts

```diff
@@ -0,0 +1,1154 @@
+                }
+            }
+            for (let i = 0; i < scopes.length; i++) {
+                const scope = scopes[i];
+                const resolvedSymbol = checker.resolveName(symbol.name, scope, symbol.flags);
+                if (resolvedSymbol === symbol) {
+                    continue;
+                }
+                if (!substitutionsPerScope[i].has(symbolId)) {
+                    const substitution = tryReplaceWithQualifiedNameOrPropertyAccess(symbol.exportSymbol || symbol, scope, isTypeName);
+                    if (substitution) {
+                        substitutionsPerScope[i].set(symbolId, substitution);
+                    }
+                    else if (isTypeName) {
+                        errorsPerScope[i].push(createDiagnosticForNode(identifier, Messages.TypeWillNotBeVisibleInTheNewScope));
+                    }
+                    else {
+                        usagesPerScope[i].usages.set(identifier.text as string, { usage, symbol, node: identifier });
+                    }
+                }
+            }
+            return symbolId;
+        }
+
+        function checkForUsedDeclarations(node: Node) {
+            // If this node is entirely within the original extraction range, we don't need to do anything.
+            if (node === targetRange.range || (isArray(targetRange.range) && targetRange.range.indexOf(node as Statement) >= 0)) {
+                return;
+            }
+
+            // Otherwise check and recurse.
+            const sym = checker.getSymbolAtLocation(node);
+            if (sym && visibleDeclarationsInExtractedRange.some(d => d === sym)) {
+                for (const scope of errorsPerScope) {
+                    scope.push(createDiagnosticForNode(node, Messages.CannotExtractExportedEntity));
+                }
+                return true;
+            }
+            else {
+                forEachChild(node, checkForUsedDeclarations);
+            }
+        }
+
+        function tryReplaceWithQualifiedNameOrPropertyAccess(symbol: Symbol, scopeDecl: Node, isTypeNode: boolean): PropertyAccessExpression | EntityName {
+            if (!symbol) {
+                return undefined;
+            }
+            if (symbol.getDeclarations().some(d => d.parent === scopeDecl)) {
+                return createIdentifier(symbol.name);
+            }
+            const prefix = tryReplaceWithQualifiedNameOrPropertyAccess(symbol.parent, scopeDecl, isTypeNode);
+            if (prefix === undefined) {
+                return undefined;
+            }
+            return isTypeNode ? createQualifiedName(<EntityName>prefix, createIdentifier(symbol.name)) : createPropertyAccess(<Expression>prefix, symbol.name);
+        }
+    }
+
+    function getParentNodeInSpan(node: Node, file: SourceFile, span: TextSpan): Node {
+        while (node) {

```

Comment:
>node [](start = 15, length = 4)

If we don't expect the argument to be undefined, the loop exit condition is probably redundant.